stan/math/rev/core/profiling.hpp

```diff
@@ -0,0 +1,179 @@
+//     msg << "Profile '" << name << "' ";
+//     msg << "already started!";
+//     throw std::runtime_error(msg.str());
+//   }
+//   if (p->second.meta.rev_pass_active) {
+//     std::ostringstream msg;
+//     msg << "Reverse pass active for a non-autodiff profile '";
+//     msg << name << "'."
+//         << "Please file a bug report!";
+//     throw std::runtime_error(msg.str());
+//   }
+//   p->second.meta.fwd_pass_start_tp = std::chrono::steady_clock::now();
+//   p->second.meta.fwd_pass_active = true;
+// }
+
+// /**
+//  * Stops profiling the forward pass for the profile
+//  * with the specified name. Does not profile the reverse
+//  * pass.
+//  * 
+//  * @tparam T profile type which must not a var
+//  * 
+//  * @param name the name of the profile to stop
+//  * @param profiles the map used for storing profiling info
+//  */
+// template <typename T, require_all_not_var_t<T>* = nullptr>
+// inline void profile_stop(const char* name, profile_map& profiles) {
+//   profile_key key = {name, std::this_thread::get_id()};
+//   profile_map::iterator p = profiles.find(key);
+//   if (p == profiles.end()) {
+//     std::ostringstream msg;
+//     msg << "Stopping a non-registered profile '" << name
+//         << "'. Please file a bug report!";
+//     throw std::runtime_error("Stopping ");
+//   }
+//   if (!p->second.meta.fwd_pass_active) {
+//     std::ostringstream msg;
+//     msg << "Stopping forward pass profile '" << name
+//         << "' that was not started. Please file a bug report!";
+//     throw std::runtime_error(msg.str());
+//   }
+//   p->second.fwd_pass
+//       += std::chrono::duration<double>(std::chrono::steady_clock::now()
+//                                        - p->second.meta.fwd_pass_start_tp)
+//              .count();
+//   p->second.meta.fwd_pass_active = false;
+// }
+
+/**
+ * Starts profiling the forward pass for the profile
+ * with the specified name and adds a var that stops
+ * profiling the reverse pass in the reverse callback.
+ * 
+ * @tparam T profile type which must not a var
+ * 
+ * @param name the name of the profile to start
+ * @param profiles the map used for storing profiling info
+ */
+template <typename T, require_all_var_t<T>* = nullptr>
+inline auto profile_start(const char* name, profile_info& profile) {
+  using std::chrono::duration;
+  using std::chrono::steady_clock;
+  if (profile.meta.fwd_pass_active) {
+    std::ostringstream msg;
+    msg << "Profile '" << name << "' ";
+    msg << "already started!";
+    throw std::runtime_error(msg.str());
+  }
+  profile.meta.fwd_pass_start_tp = steady_clock::now();
+  profile.meta.fwd_pass_active = true;
+  profile.meta.start_chain_stack_size = ChainableStack::instance_->var_stack_.size();
+  reverse_pass_callback([name, &profile]() mutable {
+    profile.rev_pass_time += duration<double>(steady_clock::now()
+                             - profile.meta.rev_pass_start_tp).count();
+    profile.meta.rev_pass_active = false;
```

Comment:
I guess it should throw an error if it's already been stopped? That might be too much error checking, cause it should never be possible.

Reply:
I had that here but then removed it because this is internal and on the API level that would mean the object was destructed twice, which cant happen. Not sure if my reasoning is valid tho.